js/viewmodel.js

```diff
@@ -0,0 +1,221 @@
+                .remove();
+            return true;
+        } catch (e) {
+            return false;
+        }
+    }
+
+    function clear() {
+        bp.entities.map(function (e) { return e.position; })
+            .forEach(clearPosition);
+    }
+
+    function onItemRemoved(callback) {
+        return events.on('itemRemoved', callback);
+    }
+
+    function onItemPlaced(callback) {
+        return events.on('itemPlaced', callback);
+    }
+
+    function clearPosition(point) {
+        var wasCleared = bp.removeEntityAtPosition(point);
+
+        if (wasCleared) {
+            console.log('cleared blueprint position', point);
+            events.emit('itemRemoved', { point: point });
+        }
+        return wasCleared;
+    }
+
+    function tryPlaceSelectedItem(point) {
+        if (!selectedItem) { return false; }
+        try {
+            var entity = bp.createEntity(
+                selectedItem.name,
+                point,
+                selectedItem.direction
+            );
+            if (entity.directionType) {
+                entity.directionType = selectedItem.type;
+            }
+            console.log('added to blueprint:', selectedItem, 'at', point, entity);
+            events.emit('itemPlaced', {
+                point: point,
+                item: Object.assign({}, selectedItem),
+                entity: entity
+            });
+            return true;
+        } catch (e) {
+            console.warn('failed to create entity', e);
+            return false;
+        }
+    }
+
+    function rotateSelectedItem() {
+        if (selectedItem && selectedItem.canRotate) {
+            console.log('rotating', selectedItem);
+            selectedItem.direction = (selectedItem.direction + 2) % 8;
+            pickPlacable(selectedItem);
+        }
+    }
+
+    function onItemSelected(callback) {
+        return events.on('itemSelected', callback);
+    }
+
+    // TODO: this name feels wrong
+    function pickPlacable(placeable) {
+        console.log('selected', placeable);
+        selectedItem = placeable && Object.assign({}, placeable);
+        events.emit('itemSelected', selectedItem);
+    }
+
+    function getPlaceableItems() {
+        var rawEntities = Blueprint.getEntityData();
+        return Object.keys(rawEntities)
+            .map(function (name) {
+                return Object.assign(
+                    {
+                        name: name
+                    },
+                    rawEntities[name],
+                    // fix bad data in factorio-blueprint
+                    name === 'steam_turbine' && { width: 5, height: 3 },
+                    name === 'boiler' && { width: 3, height: 2 },
+                    // fix spots where height/width are swapped
+                    name === 'pump' && { width: 2, height: 1 },
+                    name === 'decider_combinator' && { width: 2, height: 1 },
+                    name === 'arithmetic_combinator' && { width: 2, height: 1 }

```

Comment:
can we move these into a switch statement? It took me a while to understand what these && were doing

Reply:
will do